File: go/cli/info.go
package cli

import (
	"flag"
	"io"

	compiled "github.com/ball-lang/ball/go/cli/compiled"
)

// cmdInfo implements `ball info <program.ball.json>` (issue #570): print the
// program's structure — name/version, entry point, and a per-module breakdown.
//
// The report text is NOT written here: it comes from the self-hosted CLI core
// (dart/shared/lib/cli_core.dart's infoReport, compiled through the Ball → Go
// compiler into go/cli/compiled), so it is byte-identical to what
// `dart run dart/cli/bin/ball.dart info` prints — proven by
// cli_core_parity_test.go's golden comparison. The Go sibling of
// rust/cli/src/commands/info.rs.
//
// The program is LOADED before the report is built, so a missing file (exit 3)
// or a malformed program (exit 2) reports its own, more specific failure — the
// same ordering rust/cli's `info` uses.
func cmdInfo(args []string, w io.Writer) *cliError {
	const usage = "ball info <program.ball.json>"
	view, cerr := loadCliCoreView("info", usage, flag.NewFlagSet("info", flag.ContinueOnError), args)
	if cerr != nil {
		return cerr
	}
	return printLine(w, compiled.InfoReport(view))
}

File: python/cli/ball_cli/commands/info.py
"""``ball info <program.ball.json>`` — print a program's structure (issue #570).

The report text is not written here: it comes from the self-hosted CLI core
(``cli_core.infoReport``, compiled through the Ball → Python compiler — see
:mod:`ball_cli.cli_core`), so it is byte-identical to what the Dart CLI prints.
The Python sibling of ``rust/cli/src/commands/info.rs`` and ``go/cli/info.go``.

The program is LOADED before the CLI core is resolved, so a missing file (exit 3)
or a malformed program (exit 2) reports its own, more specific failure rather
than being masked by the "CLI core not built" message.
"""

from __future__ import annotations

from typing import TextIO

from ..argparse_util import StreamParser
from ..cli_core import program_view, reports


def command(args: list[str], stdout: TextIO, stderr: TextIO) -> int:
    parser = StreamParser(
        prog="ball info",
        description="Print a Ball program's structure (modules, types, functions).",
        out=stdout,
        err=stderr,
    )
    parser.add_argument("input", metavar="<program.ball.json>", help="path to a .ball.json program")
    ns = parser.parse_args(args)

    view = program_view(ns.input)
    stdout.write(reports().infoReport(view) + "\n")
    return 0
